### PR TITLE
[AIR/Docs] Small improvements to Train user guide

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

Train User Guide:
* Expands the Results section
* Combines reporting and checkpointing sections (both making use of `session`)
* Moving the Ray Datasets section up
* Adding some extra links and references

Also fixes typo of `Results` -> `Result` in a few places.

This is intended to be just an iterative improvement to the Train user guide. Target structure still needs to be decided on.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/user_guide.rst [hunk 1]
```diff
@@ -328,168 +328,184 @@ perform hyperparameter tuning with Ray Train, please refer to the
...
 .. TODO(ml-team) Flesh this section out.
 
 The return of a ``Trainer.fit`` is a :class:`Result` object, containing
 information about the training run. You can access it to obtain saved checkpoints,
 metrics and other relevant data.
 
-.. _train-log-dir:
+.. automodule:: ray.air.result
+    :members:
+    :noindex:
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:342]: Seems a bit weird to include this directly in the user guide - I think it's okay to remove this as long as the previous reference links to the API docs (which right now isn't happening).

File: doc/source/train/user_guide.rst [hunk 2]
```diff
@@ -328,168 +328,184 @@ perform hyperparameter tuning with Ray Train, please refer to the
...
-.. TODO link to Training Run Iterator API as a 3rd option for logging.
-
-.. _train-monitoring:
+.. code-block:: python
 
-Logging, Monitoring, and Callbacks
-----------------------------------
+    from pprint import pprint
+
+    pprint(results.metrics)
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:352]: nit: Should we rename `results` to `result` for consistency?

File: doc/source/train/user_guide.rst [hunk 3]
```diff
@@ -328,168 +328,184 @@ perform hyperparameter tuning with Ray Train, please refer to the
...
+.. automodule:: ray.air.result
+    :members:
+    :noindex:
 
-Log Directory Structure
-~~~~~~~~~~~~~~~~~~~~~~~
+For example, you can:
 
-Each ``Trainer`` will have a local directory created for logs and checkpoints.
+* print the metrics for the last iteration:
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:346]: nit 
```suggestion
* Print the metrics for the last training iteration:
```

File: doc/source/train/user_guide.rst [hunk 4]
```diff
@@ -328,168 +328,184 @@ perform hyperparameter tuning with Ray Train, please refer to the
...
-and also has a :ref:`Callback interface <train-callbacks>` to perform actions on these intermediate results (such as logging, aggregations, etc.).
-You can use either the :ref:`built-in callbacks <air-builtin-callbacks>` that Ray AIR provides,
-or implement a :ref:`custom callback <train-custom-callbacks>` for your use case. The callback API
-is shared with Ray Tune.
+.. code-block:: python
 
-Reporting intermediate results
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    print(results.metrics_dataframe)
+    #    epoch  _timestamp  _time_this_iter_s  _training_iteration  time_this_iter_s  ...        node_ip  time_since_restore  timesteps_since_restore  iterations_since_restore warmup_time
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:382]: Formatting for this is weird... can we further reduce the columns so it fits on a regularly sized page?

File: doc/source/train/user_guide.rst [hunk 5]
```diff
@@ -328,168 +328,184 @@ perform hyperparameter tuning with Ray Train, please refer to the
...
+    # 0      0  1657829301           0.002509                    1          2.420308  ...  172.31.43.110            2.420308                        0                         1    0.003495
+    # 1      1  1657829301           0.001195                    2          0.010221  ...  172.31.43.110            2.430529                        0                         2    0.003495
 
-Ray AIR provides a ``session.report(metrics, checkpoint=None)`` API for reporting intermediate
-results from the training function (run on distributed workers) up to the
-``Trainer`` (where your python script is executed).
+* Obtain the :class:`Checkpoint`, used for resuming training, prediction and serving.
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:386]: Example?

File: doc/source/train/user_guide.rst [hunk 6]
```diff
@@ -328,168 +328,184 @@ perform hyperparameter tuning with Ray Train, please refer to the
...
 
-    from ray.air import session
+Each ``Trainer`` will have a local directory created for logs and checkpoints.
 
-    def train_func():
-        ...
-        for i in range(num_epochs):
-            results = model.train(...)
-            session.report({"results": results})
+You can obtain the path to the directory by accessing the ``log_dir`` attribute
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:395]: Add a super simple code example for this?

File: doc/source/train/user_guide.rst [hunk 7]
```diff
@@ -498,10 +514,30 @@ useful for:
...
+
+.. code-block:: python
+
+    from ray.air import session
+
+    def train_func():
+        ...
+        for i in range(num_epochs):
+            results = model.train(...)
+            session.report({"results": results})
```
Comment (matthewdeng) [inline doc/source/train/user_guide.rst:535]: I feel like this is missing some important details around what happens to these reported metrics.

(Second to this, maybe we should mention that only worker 0 results are used).

**Review by richardliaw — APPROVED** [review]

**Pull request merged**